### PR TITLE
fix(aggregation): skip per-entity fan-out for local-only entities

### LINKED ISSUE #381: Per-entity endpoints fan out to all peers, emit spurious failed_peers for local entities
# Bug report

## Steps to reproduce

1. Set up an aggregating gateway A with a peer B registered (standard multi-instance / cross-ECU aggregation setup).
2. Ensure A hosts some entities locally (e.g. a component ``local-comp`` with apps) and B hosts its own entities.
3. From an HTTP client reach A and query a per-entity collection for a LOCAL entity, e.g. ``GET /components/local-comp/logs``.

## Expected behavior

A returns its local logs for ``local-comp``. Since no peer hosts ``local-comp``, no fan-out is needed and the response carries no ``partial`` / ``failed_peers`` marker:

```
{"items": [...], "x-medkit": {...}}
```

## Actual behavior

A runs the local query, then unconditionally fans out ``GET /components/local-comp/logs`` to every healthy peer. B has no such entity and returns 404. A marks the response as partial:

```
{"items": [...], "x-medkit": {"partial": true, "failed_peers": ["peer_B"]}}
```

The same happens for per-entity endpoints across logs, data, operations, configurations, health, and per-entity faults. Callers get a misleading ``partial: true`` even though nothing is actually broken - the forwarded path simply doesn't exist on the peer, because the entity is local to A.

## Root cause

``src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp::merge_peer_items`` forwards ``req.path`` verbatim to every healthy peer via ``AggregationManager::fan_out_get``. There is no check for whether the entity in the path is actually hosted by any peer. ``AggregationManager::find_peer_for_entity`` already provides this information but is not consulted here.

For per-entity paths ( ``/components/{id}``, ``/apps/{id}``, ``/areas/{id}``, ``/functions/{id}`` with a collection suffix ), if ``find_peer_for_entity(id)`` returns nullopt the entity is local-only and fan-out is guaranteed to produce a 404 from every peer.

## Expected fix

``merge_peer_items`` should extract the entity id from ``req.path`` and skip fan-out when no peer hosts that entity. Global collection endpoints (e.g. ``/faults``, ``/health``) that have no entity id in the path keep the current fan-out-to-all behavior.

## Environment

- ros2_medkit version: 0.4.0 (main @ 87937fb)
- ROS 2 distro: Jazzy
- OS: Ubuntu 24.04

### PR DESCRIPTION
**Pull request by bburda**
## Summary

Per-entity collection endpoints (`/components/{id}/logs`, `/apps/{id}/data`, etc.) no longer fan out to peers that do not host the entity. Before this change every such request produced `partial: true` with spurious `failed_peers` whenever the entity was local to the aggregating gateway.

Targeting the feat branch so the fix ships alongside the update-provider work; can be rebased onto main if merged independently.

---

## Issue

- closes #381

---

## Type

- [x] Bug fix
- [ ] New feature or tests
- [ ] Breaking change
- [ ] Documentation only

---

## Testing

- New `extract_entity_id_for_fan_out` helper parses per-entity paths (`components`, `apps`, `areas`, `functions`) and returns ``std::nullopt`` for global / detail / unknown shapes.
- `merge_peer_items` consults ``AggregationManager::find_peer_for_entity`` before fanning out; if no peer hosts the entity the fan-out is skipped.
- Added unit tests for the helper + two `merge_peer_items` behavior tests (skip local, still fan out for global endpoints).
- Updated three existing `merge_peer_items` tests to populate the routing table so the entity resolves to the mock peer; otherwise the new gate would kick in and the old tests would stop exercising the fan-out code path.
- All 23 `FanOutHelpers.*` tests pass.

Reviewer steps:
1. ``colcon build --packages-up-to ros2_medkit_gateway``
2. ``./build/ros2_medkit_gateway/test_fan_out_helpers --gtest_filter='FanOutHelpers.*'`` -> expect 23/23 PASSED.
3. In a multi-instance setup, query ``/components/<local-id>/logs`` and confirm the response no longer carries ``partial: true`` / ``failed_peers`` for peers that do not host the component.

---

## Checklist

- [x] Breaking changes are clearly described (no breaking changes)
- [x] Tests were added or updated if needed
- [ ] Docs were updated if behavior or public API changed

### CODE REVIEW

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 1]
```diff
@@ -76,6 +108,14 @@ inline void merge_peer_items(AggregationManager * agg, const httplib::Request &
...
     return;
   }
+  // For per-entity collection paths, skip fan-out when the entity is local-only
+  // (no peer hosts it). Otherwise every peer returns 404 and the response is
+  // falsely flagged as `partial` with spurious `failed_peers`. Global endpoints
+  // with no entity id in the path keep fan-out-to-all behavior.
+  if (auto entity_id = extract_entity_id_for_fan_out(req.path);
+      entity_id.has_value() && !agg->find_peer_for_entity(*entity_id).has_value()) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp]: Good catch - the `find_peer_for_entity(id).has_value() == false` gate was indeed wrong for collision-merged Areas/Functions (entity_merger.cpp:78,128 skip the routing table on ID collision) and for hierarchical parent Components (aggregation_manager.cpp:493-499 strips their routing entries during classification).

Fixed in 7e24252e by introducing `AggregationManager::has_peer_contributors(id)` which returns true for routed leaves *and* merged/hierarchical entities. gateway_node now populates a `peer_contributed_ids_` set from merged entity `contributors` after each discovery cycle, and the fan-out gate consults `has_peer_contributors` instead of `find_peer_for_entity`. Added a regression test for the merged Area case (`merge_peer_items_fans_out_for_merged_entity_without_routing_entry`) and a unit test for the new API (`has_peer_contributors_covers_routing_and_merged_entities`).

File: docs/api/rest.rst [hunk 2]
```diff
@@ -1077,6 +1078,13 @@ Without such a plugin, all endpoints return ``501 Not Implemented``.
 
    **Status values:** ``pending``, ``inProgress``, ``completed``, ``failed``
 
+   **Vendor extension ``x-medkit-phase``** (non-standard, SOVD-compatible):
+   ``none``, ``preparing``, ``prepared``, ``executing``, ``executed``,
+   ``failed``, ``deleting``. Differentiates "prepare completed" (``status``
+   ``completed`` + ``x-medkit-phase`` ``prepared``) from "execute completed"
+   (``status`` ``completed`` + ``x-medkit-phase`` ``executed``). Clients that
+   only consume the standard ``status`` field continue to work unchanged.
```
[the remark this answers is not in the document]
Comment (bburda) [reply docs/api/rest.rst:1156]: Resolved by the rebase onto current main: PR #379 (seed-Pending) landed meanwhile and already updated this section so it now reads `404 Not Found: Package is not registered`, which matches the post-seed behavior.

File: src/ros2_medkit_gateway/src/updates/update_manager.cpp [hunk 3]
```diff
@@ -99,6 +99,24 @@ tl::expected<void, UpdateError> UpdateManager::register_update(const nlohmann::j
...
+  if (metadata.contains("id") && metadata["id"].is_string()) {
+    const auto id = metadata["id"].get<std::string>();
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto & state_ptr = states_[id];
+    if (!state_ptr) {
+      state_ptr = std::make_unique<PackageState>();
+      state_ptr->status =
+          UpdateStatusInfo{UpdateStatus::Pending, UpdatePhase::None, std::nullopt, std::nullopt, std::nullopt};
+    }
+  }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/updates/update_manager.cpp:118]: Resolved by the rebase: the seed-Pending work landed through PR #379, so that hunk is no longer part of this branch. What remains on top of main is the fan-out fix plus `feat(updates): expose phase as x-medkit-phase vendor extension`, both of which fit the existing title/description.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 4]
```diff
@@ -76,8 +108,23 @@ inline void merge_peer_items(AggregationManager * agg, const httplib::Request &
...
+  // entities). Local-only entities produce an empty target list and skip
+  // fan-out entirely, avoiding spurious `partial: true` / `failed_peers`
+  // from peers that do not own the entity. Global collection endpoints
+  // (paths without an entity id) keep fan-out-to-all behavior.
+  std::optional<std::vector<std::string>> contributors_buffer;
+  const std::vector<std::string> * target_peers = nullptr;
+  if (auto entity_id = extract_entity_id_for_fan_out(req.path); entity_id.has_value()) {
+    contributors_buffer = agg->get_peer_contributors(*entity_id);
+    if (contributors_buffer->empty()) {
+      return;  // local-only: no peer hosts this entity
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:130]: **Minor - newly-discovered entity gap (round-2 confirmed).** When `get_peer_contributors` returns empty, fan-out is skipped entirely. Correct for local-only entities. But also fires for a peer entity that exists on a remote but hasn't been routed/contributed yet (between discovery cycles). During that window the client sees local-only data even though the entity actually lives on a peer.

Low risk (discovery cycles are periodic, window is brief), and the alternative (fan-out to all peers on unknown entity) brings back the spurious `partial: true` this PR fixes. Current behavior is the right trade-off - just worth a one-line note in `design/aggregation.rst`'s target-filtered fan-out section:
> "Entities not yet present in the routing or contributor tables are treated as local; their per-entity fan-out is deferred until the next discovery cycle."

File: src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp [hunk 5]
```diff
@@ -300,3 +308,338 @@ TEST(FanOutHelpers, merge_peer_items_creates_items_when_missing_and_peer_has_dat
...
+}
+
+TEST(FanOutHelpers, extract_entity_id_detail_endpoint_returns_nullopt) {
+  // `/components/<id>` (no trailing collection suffix) is routed to the owning
+  // peer via the forwarding path, not fan-out. The helper returns nullopt so
+  // the caller falls through to the current global behavior.
+  EXPECT_FALSE(extract_entity_id_for_fan_out("/api/v1/components/ecu-primary").has_value());
+  EXPECT_FALSE(extract_entity_id_for_fan_out("/api/v1/apps/helloApp").has_value());
+}
+
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp:342]: **Minor - missing edge-case tests for `extract_entity_id_for_fan_out` (round-2 confirmed code handles them, but no tests).** Three untested patterns:

1. **Trailing slash**: `/api/v1/components/ecu-a/logs/` - R2 traced: function finds entity ID correctly (next `/` after entity segment). No test.
2. **Empty entity ID**: `/api/v1/components//logs` - R2 traced: `id.empty()` guard at line 72-73 returns nullopt. No test.
3. **Query-string path**: `/api/v1/apps/id/data?limit=10` - R2 traced: httplib splits `req.path` from `req.params`, so query never reaches this function. No test proves it.

All 3 are quick additions (~5 lines each):
```cpp
EXPECT_EQ(extract_entity_id_for_fan_out("/api/v1/components/ecu-a/logs/"), "ecu-a");
EXPECT_EQ(extract_entity_id_for_fan_out("/api/v1/components//logs"), std::nullopt);
// query-string test needs httplib::Request with .path set separately
```

File: src/ros2_medkit_gateway/src/aggregation/aggregation_manager.cpp [hunk 6]
```diff
@@ -609,22 +653,35 @@ void AggregationManager::forward_request(const std::string & peer_name, const ht
...
   struct PeerResult {
     std::string peer_name;
     bool success{false};
     nlohmann::json items = nlohmann::json::array();
   };
 
+  // A non-null, empty target list means "no peers match" - skip fan-out
+  // entirely so callers don't see spurious partial flags.
+  if (target_peers != nullptr && target_peers->empty()) {
+    return FanOutResult{};
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/aggregation/aggregation_manager.cpp]: Good catch, fixed. `fan_out_get` early return now explicitly sets `merged_items = nlohmann::json::array()` so the invariant (always a JSON array, possibly empty) holds across all paths.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 7]
```diff
@@ -45,6 +48,35 @@ inline std::string url_encode_param(const std::string & value) {
...
+      // Entity detail endpoint like `/components/<id>` (no trailing collection).
+      // These are routed to peers via a different mechanism; not a fan-out case.
+      return std::nullopt;
+    }
+    std::string id = path.substr(start, end - start);
+    if (id.empty()) {
+      return std::nullopt;
+    }
+    return id;
+  }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:84]: Updated the docstring to reflect the intentional behavior. Matching is deliberately permissive - any `/{components|apps|areas|functions}/<id>/<anything>` path is treated as per-entity. Tightening to a whitelist would require bumping it on every new per-entity sub-endpoint, and the failure mode of the lax approach is benign (future paths still get target-filtered fan-out to the contributing peers, same as routed state). Explicit `std::nullopt` callouts for detail endpoints, empty id, and global paths are in the docstring + tests (`extract_entity_id_detail_endpoint_returns_nullopt`, `extract_entity_id_empty_segment_returns_nullopt`, `extract_entity_id_global_endpoints_return_nullopt`).

File: src/ros2_medkit_gateway/src/updates/update_manager.cpp [hunk 8]
```diff
@@ -157,7 +157,7 @@ tl::expected<void, UpdateError> UpdateManager::delete_update(const std::string &
     if (had_state) {
       auto it = states_.find(id);
       if (it != states_.end() && it->second) {
-        it->second->phase = UpdatePhase::Failed;
+        it->second->status.phase = UpdatePhase::Failed;
       }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/updates/update_manager.cpp:169]: Fixed. `delete_update` rollback now updates all three fields atomically: `status.status = Failed`, `status.phase = Failed`, `status.error_message = err.message`. Added `UpdateManagerFailureTest.DeleteRollbackUpdatesStatusPhaseAndErrorMessage` which asserts the serialized payload matches `{status:failed, x-medkit-phase:failed, error:<backend message>}`.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp (hunk 4 above)
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:130]: Good observation, documented. Added the deferred-fan-out note to `aggregation.rst` so the trade-off is explicit:

> Entities freshly announced on a peer but not yet reflected in the local routing/contributor tables (a brief window between discovery cycles) are treated as local-only: their per-entity fan-out is deferred until the next cycle rebuilds the tables. This is a deliberate trade-off against re-enabling the spurious `partial: true` path.

File: src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp (hunk 5 above)
Comment (bburda) [reply src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp:342]: Added all three tests:

- `FanOutHelpers.extract_entity_id_trailing_slash_still_matches` - `/components/ecu-a/logs/` -> `ecu-a`
- `FanOutHelpers.extract_entity_id_empty_segment_returns_nullopt` - `/components//logs` -> nullopt
- `FanOutHelpers.extract_entity_id_ignores_query_string_because_httplib_splits_it` - documents the httplib contract that `req.path` never contains `?`.

**Review by mfaferek93 — APPROVED** [review]
